src/DetectCyclicDependencies.cpp

```diff
@@ -0,0 +1,155 @@
+#include "DagData.hpp"
+#include "DetectCyclicDependencies.hpp"
+
+//This file is an implementation of the algorithm described in https://www.youtube.com/watch?v=rKQaZuoUR4M
+
+struct VisitState
+{
+    enum State
+    {
+        NotYetVisited,
+        BeingVisited,
+        GuaranteedNoCycles
+    };
+};
+
+struct CycleDetector
+{
+    //stores the state/color of each node
+    Buffer<VisitState::State> m_State;
+
+    //bookkeeping so that when we do find a cycle, we can backtrack our steps to figure out the exact path
+    Buffer<int32_t> m_ScheduledByWho;
+
+    int m_NodeCount;
+    MemAllocHeap* m_Heap;
+    const Frozen::Dag* m_Dag;
+
+public:
+    CycleDetector(const Frozen::Dag* dag, MemAllocHeap* heap)
+    {
+        m_Dag = dag;
+        m_Heap = heap;
+        m_NodeCount = m_Dag->m_NodeCount;
+        BufferInitWithCapacity(&m_State, m_Heap, m_NodeCount);
+        BufferInitWithCapacity(&m_ScheduledByWho, m_Heap, m_NodeCount);
+        for (int i=0; i!=m_NodeCount;i++)
+        {
+            m_State[i] = VisitState::NotYetVisited;
+            m_ScheduledByWho[i] = -1;
+        }
+    }
+
+    ~CycleDetector()
+    {
+        BufferDestroy(&m_State, m_Heap);
+        BufferDestroy(&m_ScheduledByWho, m_Heap);
+    }
+
+    bool Detect()
+    {
+        while(true)
+        {
+            int root = FindUnvisitedNode();
+            if (root == -1)
+            {
+                return false;
+            }
+            if (DepthFirstSearch(root))
+                return true;
+        }
+    }
+
+private:
+    bool DepthFirstSearch(int node)
+    {
+        auto visitDependencies = [&](const FrozenArray<int32_t>& deps) -> bool
+        {
+            for (auto dep: deps)
+            {
+                int depState = m_State[dep];
+                if (VisitState::GuaranteedNoCycles == depState)
+                    continue;
+
+                if (VisitState::BeingVisited == depState)
+                {
+                    PrintCycleFor(node, dep);
+                    return true;
+                }
+
+                m_ScheduledByWho[dep] = node;
+                if (DepthFirstSearch(dep))
+                    return true;
+            }
+            return false;
+        };
+
+        m_State[node] = VisitState::BeingVisited;
+        if (visitDependencies(m_Dag->m_DagNodes[node].m_ToBuildDependencies))
+            return true;
+        if (visitDependencies(m_Dag->m_DagNodes[node].m_ToUseDependencies))
+            return true;
+        m_State[node] = VisitState::GuaranteedNoCycles;
+        return false;
+    }
+
+    int FindUnvisitedNode()
```

Comment:
In the worst case - a very flat graph where the majority of nodes will have no dependencies - overall cost of this function will approach O(N^2).

Reply:
fixed